source/iterator/snapshot.go

```diff
@@ -0,0 +1,246 @@
+	orderingColumn string
+	// maxValue from ordering column.
+	maxValue any
+	// batchSize size of batch.
+	batchSize int
+	// position last recorded position.
+	position *position.Position
+	// columnTypes column types from table.
+	columnTypes map[string]string
+}
+
+func newSnapshotIterator(
+	ctx context.Context,
+	db *sqlx.DB,
+	table, orderingColumn string,
+	keys, columns []string,
+	batchSize int,
+	position *position.Position,
+	columnTypes map[string]string,
+) (*snapshotIterator, error) {
+	var err error
+
+	it := &snapshotIterator{
+		db:             db,
+		table:          table,
+		columns:        columns,
+		keys:           keys,
+		orderingColumn: orderingColumn,
+		batchSize:      batchSize,
+		position:       position,
+		columnTypes:    columnTypes,
+	}
+
+	err = it.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	if position != nil {
+		it.maxValue = position.SnapshotMaxValue
+	} else {
+		err = it.setMaxValue(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("set max value: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// HasNext check ability to get next record.
+func (i *snapshotIterator) HasNext(ctx context.Context) (bool, error) {
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	// check new batch.
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Next get new record.
+func (i *snapshotIterator) Next(ctx context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := i.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := coltypes.TransformRow(ctx, row, i.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[i.orderingColumn]; !ok {
+		return sdk.Record{}, ErrNoOrderingColumn
+	}
+
+	pos := position.Position{
+		IteratorType:             position.TypeSnapshot,
+		SnapshotLastProcessedVal: transformedRow[i.orderingColumn],
+		SnapshotMaxValue:         i.maxValue,
+		Time:                     time.Now(),
+	}
+
+	sdkPos, err := pos.ConvertToSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("convert position %w", err)
+	}
+
+	keysMap := make(map[string]any)
+	for _, val := range i.keys {
+		if _, ok := transformedRow[val]; !ok {
+			return sdk.Record{}, fmt.Errorf("key %v, %w", val, ErrNoKey)
+		}
+
+		keysMap[val] = transformedRow[val]
+	}
+
+	transformedRowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	i.position = &pos
+
+	metadata := sdk.Metadata(map[string]string{metadataTable: i.table})
+	metadata.SetCreatedAt(time.Now())
+
+	return sdk.Util.Source.NewRecordSnapshot(
+			sdkPos,
+			metadata,
+			sdk.StructuredData(keysMap),
+			sdk.RawData(transformedRowBytes)),
```

Comment:
As I remember we prefer use RawData because Conduit has proto issue with time types something like this:
```
proto: invalid type: time.Time" component=pipeline.Service node_id=068b4eb5-0b96-416c-8975-ddde051a11e6
```